File: backend/src/api/components/schemas.ts
import { z } from "@hono/zod-openapi"

export const Language = z
  .object({
    name: z.string(),
    version: z.string(),
  })
  .openapi("Language")

export const TestCase = z
  .object({
    input: z.string(),
    output: z.string(),
  })
  .openapi("TestCase")

export const Problem = z
  .object({
    body: z.string(),
    id: z.number().int().nonnegative(),
    inputConditions: z.string().nullable(),
    outputConditions: z.string().nullable(),
    supported_languages: z.array(Language),
    test_cases: z.array(TestCase),
    title: z.string(),
    usePropertyBasedTesting: z.boolean(),
  })
  .openapi("Problem")

export const ProblemCreate = Problem.omit({ id: true }).openapi("ProblemCreate")

export const ProblemUpdate = Problem.omit({ id: true }).openapi("ProblemUpdate")

export const SubmissionStatus = z
  .enum(["Accepted", "WrongAnswer", "RuntimeError", "CompileError"])
  .openapi("SubmissionStatus")

export const TestStatus = z.enum(["Passed", "Failed"]).openapi("TestStatus")

export const SubmissionResult = z
  .object({
    message: z.string().optional(),
    status: SubmissionStatus,
  })
  .openapi("SubmissionResult")

export const TestResult = z
  .object({
    message: z.string().optional(),
    status: TestStatus,
    test_case_id: z.number().int().nonnegative(),
  })
  .openapi("TestResult")

export const Submission = z
  .object({
    code: z.string(),
    id: z.number().int().nonnegative(),
    language: Language,
    problem_id: z.number().int().nonnegative(),
    result: SubmissionResult,
    student_id: z.number().int().nonnegative(),
    submitted_at: z.string().datetime(),
    test_results: z.array(TestResult),
  })
  .openapi("Submission")

export const SubmissionCreate = Submission.omit({
  id: true,
  problem_id: true,
  result: true,
  student_id: true,
  submitted_at: true,
  test_results: true,
}).openapi("SubmissionCreate")

File: backend/src/api/components/schemas.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct Problem {
    id: u64,
    title: String,
    body: String,
    test_cases: Vec<TestCase>,
    supported_languages: Vec<Language>,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct ProblemCreate {
    title: String,
    body: String,
    test_cases: Vec<TestCase>,
    supported_languages: Vec<Language>,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct ProblemUpdate {
    title: Option<String>,
    body: Option<String>,
    test_cases: Option<Vec<TestCase>>,
    supported_languages: Option<Vec<Language>>,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct TestCase {
    input: String,
    output: String,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct Language {
    name: String,
    version: String,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct Submission {
    id: u64,
    problem_id: u64,
    student_id: u64,
    code: String,
    language: Language,
    result: SubmissionResult,
    test_results: Vec<TestResult>,
    submitted_at: DateTime<Utc>,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct SubmissionCreate {
    problem_id: u64,
    student_id: u64,
    code: String,
    language: Language,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct SubmissionResult {
    status: SubmissionStatus,
    message: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub struct TestResult {
    test_case_id: u64,
    status: TestStatus,
    message: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub enum SubmissionStatus {
    #[default]
    Accepted,
    WrongAnswer,
    RuntimeError,
    CompileError,
}

#[derive(Serialize, Deserialize, Clone, ToSchema, Default)]
pub enum TestStatus {
    #[default]
    Passed,
    Failed,
}
